### ISSUE TITLE
Middle mouse button copy and paste under X11 not working

### ISSUE THREAD

**Issue by LongHairedHacker** [lapce/lapce#834]
One convenient feature under X11 is that I can select text in one application (without any explicit copy command) and paste it into another application by clicking the middle mouse button.
This uses the so called "PRIMARY" "[selection](https://www.x.org/releases/X11R7.6/doc/xorg-docs/specs/ICCCM/icccm.html#peer_to_peer_communication_by_means_of_selections)"  in contrast to the "CLIPBOARD" "selection" (I've put selection in quotes here, as it refers to the selections from the X11 Inter-Client Communication Conventions Manual) used for the explicit crtl-c ctr-v copy and paste.

Currently you can neither copy text from, nor paste text into lapce using this mechanism.

https://specifications.freedesktop.org/clipboards-spec/clipboards-latest.txt has some more details on the commonly used behaviour:
 - selecting but with no explicit copy should only set PRIMARY, never CLIPBOARD
 - middle mouse button should paste PRIMARY, never CLIPBOARD
 - explicit cut/copy commands (i.e. menu items, toolbar buttons) should always set CLIPBOARD to the currently-selected data (i.e. conceptually copy PRIMARY to CLIPBOARD)
 - explicit cut/copy commands should always set both CLIPBOARD and PRIMARY, even when copying doesn't involve a selection (e.g. a "copy url" -option which explicitly copies an url without the url being selected first)
 - explicit paste commands should paste CLIPBOARD, not PRIMARY
 - a selection becoming unselected should never unset PRIMARY

I'm not sure if this is issue with lapce itself, or rather with druid or druids X11 backend.

**Comment by lisael** [comment]
I'm not sure if it's a feature or a bug. An X11 app that doesn't honor this feature is buggy to me. I just can't use it. I'm tempted to implement this myself in PR, as I'm frustrated I can't use lapce for such a detail.

**Comment by floriankramer** [comment]
@lisael did you start implementing this? Otherwise I'd give it a go myself, as it is a feature I'm sorely missing (and it shouldn't be too hard to do, druid has direct support for getting a `Clipboard` instance which accesses the primary buffer).

**Comment by lisael** [comment]
@floriankramer do it! I've just started thinking about it and reading code and doc. You seem to have a much clearer plan than I do. Thanks!

**Comment by floriankramer** [comment]
So it turns out that druid hasn't had a new version released since Jan 10, 2021. The pr we need to do this properly is https://github.com/linebender/druid/pull/1867 which is from July of 2021. lapce uses its own fork of druid, which is slightly newer than version 0.7 but doesn't contain the linux ApplicationExt stuff.

There is likely a hacky way to get clipboard support for the editor window and terminal using druid 0.7 by manually creating a new clipboard instance, but that doesn't sound like a good idea. Instead we should try to get lapce/druid updated to include primary buffer support.

**Comment by panekj** [comment]
We use GTK, not X11

**Comment by floriankramer** [comment]
Right, that pr was for X11 (there's also a gtk one). Figured out by now that the druid `opengl_shell` branch has all the prs needed, I just looked at the wrong branch.

**Comment by floriankramer** [comment]
Got a first draft for a pr to add primary buffer support. So far it's editor window only, and kinda ugly in when it puts data into the primary buffer (after every keydown or mouseup event).
